Add custom tabs headers article (#8946)
google/WebFundamentals: src/content/en/android/custom-tabs/headers/index.md

project_path: /web/android/_project.yaml
book_path: /web/android/_book.yaml
description: Guide for adding HTTP CORS Headers in Custom Tab Intents.

{# wf_published_on: 2020-08-12 #}
{# wf_updated_on: 2020-08-12 #}
{# wf_blink_components: N/A #}

> # HTTP Request Headers in Custom Tab Intents {: .page-title }

Reviewer: Would something like this be more clear?

```suggestion
# How to add extra HTTP Request Headers to Custom Tab Intents {: .page-title }
```

## Introduction

> HTTP requests contain headers such as User-Agent or Content-Type. Apart from headers attached by browsers, Android apps may add extra headers, like Cookie or Referrer through the [`EXTRA_HEADERS`][1] Intent extra. For security reasons, Chrome filters some of the extra headers depending on how and where an intent is launched.

Reviewer: nit: You want to break text into lines with 100 chars.

> [Cross-origin][2] requests require an additional layer of security as the client and server are not owned by the same party. This guide discusses launching such requests through Chrome [custom tabs][3]. Until Chrome 83, developers could add any headers when launching a Custom Tab. From version 83, Chrome started filtering all except [whitelisted][4] cross-origin headers, since non-whitelisted headers posed a security risk. Starting with Chrome 86, it is possible to attach non-whitelisted headers to cross-origin requests, when the server and client are related using a [digital asset link][5]. This behaviour is summarised in [Table 1](#table-1).

Reviewer: I'm wondering if it's clear for developers that an Android app opening an Intent to open a URL in the browser is also a cross-origin request. Should we call this out explicitly here?

Reviewer: We're doing an effort to avoid potentially offensive terms to our docs as well. "whitelist" is one of those terms. Maybe we can replace those with "safelist" or "allowlist"?

<a name="table-1"></a>

**Chrome version** | **CORS headers allowed**
---------------- | ----------------------
before v83       | whitelisted, non-whitelisted
v83 to v85       | whitelisted
v86 onwards      | whitelisted, non-whitelisted when a digital asset link is set up

**Table 1.:** Filtering of non-whitelisted CORS headers.

> This guide shows how to set up a verified connection between the server and client and use that to send non-whitelisted http headers. You can skip to [Adding Extra Headers to CustomTab Intents for the code](#adding-extra-headers). 

Reviewer: I think the guide is showing how to add both safelisted and non-safelisted.

> ## Backgroud

Reviewer: ```suggestion
## Background
```

content-language | describes language intended for the current audience
content-type     | indicates the media type of the resource

**Table 2.:** Example whitelisted CORS headers.

The whitelisted headers are considered safe because they don’t contain sensitive user information and are unlikely to cause the server to perform potentially damaging operations.

Examples of non-whitelisted headers are shown in [Table 3](#table-3).

<a name="table-3"></a>

**Header**   | **Description**
-------------| ----------------------
bearer-token | authenticates client at a server
origin       | indicates origin of request
cookie       | contains cookies set by server

**Table 3.:** Example non-whitelisted CORS headers.

Attaching non-whitelisted headers to CORS requests is discouraged by the HTML standard and servers assume that cross-origin requests contain only whitelisted headers. Sending non-whitelisted headers from cross-origin domains would allow malicious third-party apps to craft headers that misuse user cookies that Chrome (or another browser) stores and attaches to requests. The cookies could authenticate malicious server transactions that would otherwise not be possible.

> ### Attaching Headers to Custom Tabs CORS Requests

Reviewer: I think the request themselves are not CORS. The point here seems to be how to attach CORS white listed headers. WDYT?

```suggestion
### Attaching CORS whitelisted headers to Custom Tabs requests
```

> [Custom Tabs][8] are a special way of launching web pages in a customised Chrome tab. Custom Tab intents can be created using `CustomTabsIntent.Builder()`. Headers can also be attached to these intents using a Bundle with the [`Borwser.EXTRA_HEADERS flag`][9]:

Reviewer: ```suggestion
[Custom Tabs][8] are a special way of launching web pages in a customised browser tab. Custom Tab intents can be created using `CustomTabsIntent.Builder()`. Headers can also be attached to these intents using a Bundle with the [`Borwser.EXTRA_HEADERS flag`][9]:
```

```Java
CustomTabsIntent intent = new CustomTabsIntent.Builder(session).build();

Bundle headers = new Bundle();
headers.putString("bearer-token", "Some token");
headers.putString("redirect-url", "Some redirect url");   intent.intent.putExtra(Browser.EXTRA_HEADERS, headers);

intent.launchUrl(Activity.this, Uri.parse("http://www.google.com"));
```

> Whitelisted headers can always be attached to custom tabs CORS requests. However, Chrome filters non-whitelisted headers by default. 

Reviewer: What about other browsers? Maybe we could mention that other browsers may have different behaviour but, overall, developers should expect non-whitelisted headers to be blocked.

> The supported way of including non-whitelisted headers in custom tabs is to first verify the cross-origin connection using a digital access link. The next section shows how to set these up and launch a custom tab intents with the required headers. 

Reviewer: ```suggestion
The supported way of including non-whitelisted headers in custom tabs is to first verify the cross-origin connection using a digital access link. The next section shows how to set these up and launch a Custom Tabs intent with the required headers. 
```

> ## Adding Extra Headers to CustomTab Intents {: #adding-extra-headers }

Reviewer: Maybe we need to qualify the difference between this header and the previous:
```suggestion
## Adding non-whitelisted Extra Headers to CustomTab Intents {: #adding-extra-headers }
```

### Set up digital asset links
To allow non-whitelisted headers to be passed through custom tab intents, it is necessary to set up a digital asset link between the android and web application that verifies that the author owns both applications.

Follow the [official guide][10] to set up a digital asset link. For the link relation use `"delegate_permission/common.use_as_origin"` which indicates that both apps belong to the same origin once the link is verified.

### Create custom tab intent with extra headers

> There are multiple ways for creating [custom tab intents][11]. You can use the builder available in androidX by adding the library to the build dependencies:

Reviewer: ```suggestion
There are multiple ways for creating  a [Custom Tabs intent][11]. You can use the builder available in androidX by adding the library to the build dependencies:
```

content-language | describes language intended for the current audience
content-type     | indicates the media type of the resource

**Table 2.:** Example whitelisted CORS headers.

The whitelisted headers are considered safe because they don’t contain sensitive 
user information and are unlikely to cause the server to perform potentially damaging operations.

Examples of non-whitelisted headers are shown in the following table:

**Header**   | **Description**
-------------| ----------------------
bearer-token | authenticates client at a server
origin       | indicates origin of request
cookie       | contains cookies set by server

**Table 3.:** Example non-whitelisted CORS headers.

Attaching non-whitelisted headers to CORS requests is discouraged by the HTML standard and servers 
assume that cross-origin requests contain only whitelisted headers. Sending non-whitelisted headers
from cross-origin domains would allow malicious third-party apps to craft headers that misuse user
cookies that Chrome (or another browser) stores and attaches to requests. The cookies could 
authenticate malicious server transactions that would otherwise not be possible.

> ### Attaching CORS whitelisted headers to Custom Tabs requests

Reviewer 2: ```suggestion
### Attaching CORS Whitelisted Headers to Custom Tabs Requests
```

Bundle headers = new Bundle();
headers.putString("bearer-token", "Some token");
headers.putString("redirect-url", "Some redirect url");   
intent.intent.putExtra(Browser.EXTRA_HEADERS, headers);

intent.launchUrl(Activity.this, Uri.parse("http://www.google.com"));
```

Whitelisted headers can always be attached to custom tabs CORS requests. However, Chrome filters 
non-whitelisted headers by default. Although other browsers may have different behaviour, 
developers should expect non-whitelisted headers to be blocked in general.

The supported way of including non-whitelisted headers in custom tabs is to first verify the
cross-origin connection using a digital access link. The next section shows how to set these
up and launch a Custom Tabs intent with the required headers. 

## Adding Extra Headers to CustomTab Intents {: #adding-extra-headers }

### Set up digital asset links
To allow non-whitelisted headers to be passed through custom tab intents, it is necessary to set
up a digital asset link between the android and web application that verifies that the author
owns both applications.

Follow the [official guide][10] to set up a digital asset link. For the link relation use 

> "delegate_permission/common.use_as_origin"` which indicates that both apps belong to the same 

Reviewer 2: ```suggestion
`"delegate_permission/common.use_as_origin"` which indicates that both apps belong to the same 
```

origin once the link is verified.

> ### Create custom tab intent with extra headers

Reviewer 2: ```suggestion
### Create Custom Tab Intent with Extra Headers
```

> There are multiple ways for creating a [Custom Tabs intent][11]. You can use the builder available

Reviewer 2: ```suggestion
There are multiple ways for creating a [custom tab intent][11]. You can use the builder available
```

in androidX by adding the library to the build dependencies:

```gradle
implementation 'androidx.browser:browser:1.2.0'
```

Build the intent and add extra headers:

```java
CustomTabsIntent constructExtraHeadersIntent(CustomTabsSession session) {
    CustomTabsIntent intent = new CustomTabsIntent.Builder(session).build();

    // Example non-cors-whitelisted headers.
    Bundle headers = new Bundle();
    headers.putString("bearer-token", "Some token");
    headers.putString("redirect-url", "Some redirect url");
    intent.intent.putExtra(Browser.EXTRA_HEADERS, headers);
    return intent;
}
```

> ### Set up a custom tabs connection to validate the asset link

Reviewer 2: ```suggestion
### Set up a Custom Tabs Connection to Validate the Asset Link
```

Reviewer 2: This and the next code example need some kind of introduction. It may seem redundant, but please provide one sentence explaining what the example does. Optionally describe any pitfalls the reader may run into when using this code or describe anything that might seem obscure. (Hint: if there's something you had trouble with when you were writing this, the reader might have trouble with it too.)

```java
// Set up a connection that warms up and validates a session.
CustomTabsServiceConnection connection = new CustomTabsServiceConnection() {
    @Override
    public void onCustomTabsServiceConnected(@NonNull ComponentName name, 
        @NonNull CustomTabsClient client) {
        // Create session after service connected.
        mSession = client.newSession(callback);
        client.warmup(0);
        // Validate the session as the same origin to allow cross origin headers.
        mSession.validateRelationship(CustomTabsService.RELATION_USE_AS_ORIGIN, 
            Uri.parse(url), null);
    }
    @Override
    public void onServiceDisconnected(ComponentName componentName) { }
};
```

> ### Set up a callback that launches the intent after validation

Reviewer 2: ```suggestion
### Set up a Callback that Launches the Intent after Validation
```

```java
// Set up a callback that launches the intent after session validated.
CustomTabsCallback callback = new CustomTabsCallback() {
    @Override
    public void onRelationshipValidationResult(int relation, @NonNull Uri requestedOrigin, 
        boolean result, @Nullable Bundle extras) {
        // Launch custom tabs intent after session was validated as the same origin.
        CustomTabsIntent intent = constructExtraHeadersIntent(mSession);
        intent.launchUrl(MainActivity.this, Uri.parse(url));
    }
};
```

### Bind the custom tabs service connection

Binding the service launches the service and the connection’s `onCustomTabsServiceConnected()`
will be called eventually. Don’t forget to unbind the service appropriately. Binding and unbinding
is commonly done in the `onStart()` and `onStop()` activity lifecycle methods.

```java
// Bind the custom tabs service connection.
// Call this in onStart()
CustomTabsClient.bindCustomTabsService(this,
    CustomTabsClient.getPackageName(MainActivity.this, null), connection);

// …
// Unbind the custom tabs service.
// Call this in onStop().
unbindService(connection);
```

### Demo application code

> More details about Custom Tabs Service can be found [here][12]. A working example app can be found

Reviewer 2: ```suggestion
You can find more details about Custom Tabs Service [here][12]. A working example app can be found
```

Reviewer 2: Constructions like 'can be found' and 'will be done' are not considered the best possible writing in English. Replace them with phrases like 'You can find' and 'the system will'. I won't hold up the article, but if you feel inclined, it would improve the writing.

{# wf_published_on: 2020-08-12 #}
{# wf_updated_on: 2020-08-12 #}
{# wf_blink_components: N/A #}

# How to add extra HTTP Request Headers to Custom Tab Intents {: .page-title }

## Introduction

HTTP requests contain headers such as User-Agent or Content-Type. Apart from headers attached by
browsers, Android apps may add extra headers, like Cookie or Referrer through the 
[`EXTRA_HEADERS`][1] Intent extra. For security reasons, Chrome filters some of the extra headers
depending on how and where an intent is launched.

[Cross-origin][2] requests require an additional layer of security as the client and server are
not owned by the same party. This guide discusses launching such requests through Chrome 
[custom tabs][3], i.e. intents launched from apps that open a URL in the browser tab. Until Chrome
83, developers could add any headers when launching a Custom Tab. From version 83, Chrome started 
filtering all except [whitelisted][4] cross-origin headers, since non-whitelisted headers posed
a security risk. Starting with Chrome 86, it is possible to attach non-whitelisted headers to 
cross-origin requests, when the server and client are related using a [digital asset link][5]. 

> This behaviour is summarised in [Table 1](#table-1).

Reviewer 3: I'm not sure it's particularly necessary to have a link to the table that appears just after the link. (Are you linking to table-1 from anywhere else? If not we can probably do away with the anchor.)

    return intent;
}
```
### Set up a Custom Tabs Connection to Validate the Asset Link

A Custom Tabs connection is used for setting up a `CustomTabsSession` between the app and the

> Chrome tab. We need the session to verify the app nad web app belong to the same origin.

Reviewer 2: ```suggestion
Chrome tab. We need the session to verify that the app and web app belong to the same origin.
```
